include/circt/Dialect/Moore/MooreOps.td

```diff
@@ -204,4 +198,446 @@ def NonBlockingAssignOp : AssignOpBase<"nonblocking_assign"> {
   }];
 }
 
+//===----------------------------------------------------------------------===//
+// Expressions
+//===----------------------------------------------------------------------===//
+
+def ConstantOp : MooreOp<"constant", [Pure]> {
+  let summary = "A constant integer value";
```

Comment:
Yes, that would be my approach. Define a new attribute that has two `APInt`s internally to capture which of the four values it is. And then possibly funnel all SV constants through that new attribute. Or allow mixing of `IntegerAttr` and `SVIntegerAttr`, but that would make folders and canonicalizers more annoying to write.

Reply:
If you need expertise on the ways it _shouldn't_ be done, I'm your man. 😉

Reply:
@fabianschuiki  - I have an implementation that I've recently finished to handle Z and X and it does indeed rely on APInt as well as the slang SVInt class. I'm happy to take a stab at that when my time permits.

Reply:
I hadn't considered using `SVInt`. That might actually be a really good approach. Just put that into attribute storage and intern it in the context.

Reply:
Ah but that would expose Slang types in CIRCT, outside the confines of `ImportVerilog`. One of the design goals is to to have all Slang types and specifics hidden away in that library, and not have it leak out into other parts of CIRCT. So Moore will want its own version of this. Could be just two `APInt`s put together, with all the interesting operations defined on them. And then an attribute that wraps that into an interned `SVIntegerAttr`.

Reply:
At this point I feel I cannot avoid pointing out my misadventures of last summer: [APLogic and APLogicConstant](https://github.com/fzi-hielscher/circt/commit/cc122db1ffdea47da95b35598d556533befb4774#diff-57476afe80d18cca13664028bf9a8fe92477418ef9c78eba50c8091ee0bbde73) . `APLogic` is/was effectively a nine-valued variant of `APInt` using 4 bits per digit. `APLogicConstant` was intended as an immutable container that picked its encoding depending on the stored value, similar to slang's `SVInt`. In the two-valued case it was, regarding its structure, an immutable `APInt` with the encoding information tucked away in the padding.
I ended up shelving them. The switching between 1 Bit, 2 Bit and 4 Bit encoding made the implementation of the operations far too complex for my liking. And I felt general arithmetic on arbitrary strings of nine-valued logic is too irrelevant in practice to be worth adding the complexity for. 
I'd be happy if something useful could rise out of its smoking ruins. But on the other hand, wouldn't it be possible to just ask if we can re-license `SVInt` and simply copy it?

For my Many Valued Logic Dialect shenanigans I have since switched to a [completely different approach](https://github.com/fzi-hielscher/circt/blob/hwlogic-rlelogic-github/include/circt/Dialect/HW/HWRLELogic.h), that does not even try to be useful for storing integers. I have implemented some basic folds with that one, but they are not on github at the moment. Since it is also made for nine valued logic and with parameterized width types in mind, it probably is not the right choice for Moore.

Reply:
I've been meaning to ask about the old PR every time I do a round of reviewing.  Should we close it?

Reply:
Some possibilities:
* different attribute kinds for simple v.s. complex contants (doesn't have to affect parsing/printing).
* different constant ops for the different cases (make the simple case simple, esp since lots of code probably at first will just not touch complex constants).

I personally still lean toward different types, the only hesitation is that parameterized are different enough that maybe any non-simple type should just go through whatever handles that.

Reply:
We could try to build this out here in the Moore dialect and for Moore folds, where every constant would be complex (to prevent folders/canonicalizers from dealing with two overlapping attribute types). And then see how we can pull out the 4-valued and 9-valued flavors into the HW dialect as a core CIRCT type.